### PR TITLE
Initial attempt at mapping CIQUAL categories to ingredients

### PR DESCRIPTION
**Pull request by teolemon**
### What
Initial attempt at mapping CIQUAL categories to ingredients.

### Why
- Agribalyse categories being CIQUAL categories, this should pave the way for refined computations.
- We should also be able to recompute approximate nutritional values from the nutritional values of each ingredient, multiplied by the CIQUAL nutritional values.

### Part of
- https://github.com/openfoodfacts/offf/issues/14

### CODE REVIEW

**Review by stephanegigandet — APPROVED** [review]

File: taxonomies/ingredients.txt [hunk 1]
```diff
@@ -9923,12 +9987,15 @@ zh:埃德姆芝士
 wikidata:en:Q597473
 wikipedia:en:https://en.wikipedia.org/wiki/Edam_cheese
 # fr:edam has 206 products in 11 languages @2018-10-28
+ciqual_food_code:en:12729
+ciqual_food_name:en:Edam cheese, from cow's milk
+ciqual_food_name:fr:Edam
 
 
-# EMMENTAL (Emmentaler or Emmenthal) is a yellow, medium-hard Swiss cheese that originated in the area around Emmental, Canton Bern.
+#descripton:en:EMMENTAL (Emmentaler or Emmenthal) is a yellow, medium-hard Swiss cheese that originated in the area around Emmental, Canton Bern.
```
Comment (teolemon) [inline taxonomies/ingredients.txt:9995]:
```suggestion
#description:en:EMMENTAL (Emmentaler or Emmenthal) is a yellow, medium-hard Swiss cheese that originated in the area around Emmental, Canton Bern.
```

File: taxonomies/ingredients.txt [hunk 2]
```diff
@@ -10608,14 +10728,48 @@ zh:马斯卡邦尼奶酪
...
 es:queso mimolette
-
-
-# MONTEREY JACK, sometimes shortened to Jack, is an American white, semi-hard cheese made using cow's milk. It is noted for its mild flavor and slight sweetness.
+ciqual_food_code:en:12740
+ciqual_food_name:en:Mimolette cheese, from cow's milk
+ciqual_food_name:fr:Mimolette, sans précision
+
+<en:mimolette
+en:Mimolette cheese young from cow's milk
```
Comment (teolemon) [inline taxonomies/ingredients.txt:10745]:
```suggestion
en:young Mimolette cheese from cow's milk
```

File: taxonomies/ingredients.txt [hunk 3]
```diff
@@ -10608,14 +10728,48 @@ zh:马斯卡邦尼奶酪
...
+fr:Mimolette jeune
+ciqual_food_code:en:12735
+ciqual_food_name:en:Mimolette cheese, young, from cow's milk
+ciqual_food_name:fr:Mimolette jeune
+
+<en:mimolette
+en:Mimolette cheese half-old from cow's milk
```
Comment (teolemon) [inline taxonomies/ingredients.txt:10752]:
```suggestion
en:half-old Mimolette cheese from cow's milk
```

File: taxonomies/ingredients.txt [hunk 4]
```diff
@@ -10608,14 +10728,48 @@ zh:马斯卡邦尼奶酪
...
+fr:Mimolette demi-vieille
+ciqual_food_code:en:12737
+ciqual_food_name:en:Mimolette cheese, half-old, from cow's milk
+ciqual_food_name:fr:Mimolette demi-vieille
+
+<en:mimolette
+en:Mimolette cheese old from cow's milk
```
Comment (teolemon) [inline taxonomies/ingredients.txt:10759]:
```suggestion
en:old Mimolette cheese from cow's milk
```

File: taxonomies/ingredients.txt [hunk 5]
```diff
@@ -10608,14 +10728,48 @@ zh:马斯卡邦尼奶酪
...
+fr:Mimolette vieille
+ciqual_food_code:en:12738
+ciqual_food_name:en:Mimolette cheese, old, from cow's milk
+ciqual_food_name:fr:Mimolette vieille
+
+<en:mimolette
+en:Mimolette cheese extra old from cow's milk
```
Comment (teolemon) [inline taxonomies/ingredients.txt:10766]:
```suggestion
en:extra old Mimolette cheese from cow's milk
```

File: taxonomies/ingredients.txt [hunk 6]
```diff
@@ -13819,14 +14011,57 @@ nl:gehakt rundvlees
...
-fr:langue de bœuf, langue de boeuf
+fr:langue de bœuf, langue de boeuf, Langue de boeuf crue
 hu:marhanyelv
 pt:Língua de vaca
+ciqual_food_code:en:40202
+ciqual_food_name:en:Tongue, beef, raw
+ciqual_food_name:fr:Langue, boeuf, crue
+
+<en:beef tongue
+en:cooked beef tongue
```
Comment (teolemon) [inline taxonomies/ingredients.txt:14027]:
```suggestion
en:cooked beef tongue
fr:langue de boeuf cuite
```

File: taxonomies/ingredients.txt [hunk 7]
```diff
@@ -14531,6 +15003,14 @@ uk:Мергези
 wikipedia:en:https://en.wikipedia.org/wiki/Merguez
 # ingredient/fr:merguez has 63 products in french @2019-03–22
 
+<en:merguez
+en:Merguez sausage raw
```
Comment (teolemon) [inline taxonomies/ingredients.txt:15007]:
```suggestion
en:Merguez sausage
```

File: taxonomies/ingredients.txt [hunk 8]
```diff
@@ -33265,6 +34135,15 @@ zh:釀酒酵母
 # zh-hk:釀酒酵母
 # zh-tw:釀酒酵母
 wikipedia:en:https://en.wikipedia.org/wiki/Baker%27s_yeast
+ciqual_food_code:en:11045
+ciqual_food_name:en:Baker's yeast, dehydrated
+ciqual_food_name:fr:Levure de boulanger, déshydratée
+
+en:Baker's yeast compressed
```
Comment (teolemon) [inline taxonomies/ingredients.txt:34142]:
```suggestion
en:compressed baker's yeast
```

File: taxonomies/ingredients.txt [hunk 9]
```diff
@@ -37633,6 +38547,13 @@ wikipedia:en:https://en.wikipedia.org/wiki/Lime_(fruit)
 carbon_footprint_fr_foodges_ingredient:fr:Citron
 carbon_footprint_fr_foodges_value:fr:0.9
 
+<en:lime
+en:Raw lime pulp
+fr:Pulpe de citron vert cru
```
Comment (teolemon) [inline taxonomies/ingredients.txt:38552]:
```suggestion
en:lime pulp
fr:Pulpe de citron vert
```

File: taxonomies/ingredients.txt [hunk 10]
```diff
@@ -46849,8 +47881,17 @@ zh:欧防风
 # zh-tw:歐防風
 wikidata:en:Q188614
 wikipedia:en:https://en.wikipedia.org/wiki/Parsnip
+ciqual_food_code:en:20181
+ciqual_food_name:en:Parsnip, raw
+ciqual_food_name:fr:Panais, cru
 # ingredient/fr:panais has 372 products in 5 languages @2018-12-4
 
+en:Cooked parsnip
```
Comment (teolemon) [inline taxonomies/ingredients.txt:47889]:
```suggestion
<fr:oignon semoule
en:Cooked parsnip
```

File: taxonomies/ingredients.txt [hunk 11]
```diff
@@ -47310,8 +48366,23 @@ zh:黑婆罗门参
...
 # ingredient/fr:salsifis has 58 products in 5 languages @2019-04-05
 
+<en:salsify
+en:Raw black salsify
+fr:Salsifis noirs crus
+ciqual_food_code:en:20197
+ciqual_food_name:en:Salsify, black, raw
+ciqual_food_name:fr:Salsifis noir, cru
+
+en:Cooked salsify
```
Comment (teolemon) [inline taxonomies/ingredients.txt:48381]:
```suggestion
<en:salsify
en:Cooked salsify
```

File: taxonomies/ingredients.txt [hunk 12]
```diff
@@ -49130,6 +50226,15 @@ fi:lanttu
 fr:rutabaga
 nb:kålrot
 sv:kålrot
+ciqual_food_code:en:20201
+ciqual_food_name:en:Rutabaga or Swede, raw
+ciqual_food_name:fr:Rutabaga, cru
+
+en:Cooked rutabaga, Cooked swede
```
Comment (teolemon) [inline taxonomies/ingredients.txt:50233]:
```suggestion
<fr:rutabaga
en:Cooked rutabaga, Cooked swede
```

File: taxonomies/ingredients.txt [hunk 13]
```diff
@@ -63280,6 +64595,27 @@ wikidata:en:Q143359
...
+ciqual_food_name:en:Capelin, raw
+ciqual_food_name:fr:Capelan, cru
+
+<en:fish
+en:carp
+fr:carpe
+nl:karper
+
+<en:carp
+en:Carp raw farmed
```
Comment (teolemon) [inline taxonomies/ingredients.txt:64611]:
```suggestion
en:farmed carp, carp from farming
```

File: taxonomies/ingredients.txt [hunk 14]
```diff
@@ -64227,21 +65653,48 @@ wikipedia:en:https://en.wikipedia.org/wiki/Salmon_as_food
...
 en:raw salmon
 de:Roher Lachs
 fr:saumon cru
 nl:rauwe zalm
 pt:Salmão cru
+# ingredient/raw-salmon has 100 products in 3 languages @2021-08-15
+
+<en:raw salmon
+en:Salmon raw farmed
+fr:Saumon cru élevage
```
Comment (teolemon) [inline taxonomies/ingredients.txt:65666]:
```suggestion
en:Raw salmon from farming, Farmed raw salmon
fr:Saumon cru d'élevage
```

File: taxonomies/ingredients.txt [hunk 15]
```diff
@@ -64227,21 +65653,48 @@ wikipedia:en:https://en.wikipedia.org/wiki/Salmon_as_food
...
+ciqual_food_code:en:26036
+ciqual_food_name:en:Salmon, raw, farmed
+ciqual_food_name:fr:Saumon, cru, élevage
+
+<en:raw salmon
+en:Salmon raw wild
+fr:Saumon cru sauvage
```
Comment (teolemon) [inline taxonomies/ingredients.txt:65673]:
```suggestion
<en:raw salmon
en:raw wild salmon
fr:Saumon cru sauvage
```

File: taxonomies/ingredients.txt [hunk 16]
```diff
@@ -64948,9 +66430,19 @@ zh:鲣
...
-wikipedia:en:https://en.wikipedia.org/wiki/Skipjack_tuna
+wikipedia:en:https://en.wikipedia.org/wiki/Skipjack_tuna 
+ciqual_food_code:en:26101
+ciqual_food_name:en:Bonito, raw
+ciqual_food_name:fr:Bonite, crue
 # ingredient/fr:thon-listao has 138 products @2019-01-02
 
+<en:skipjack tuna
+en:Bonito raw
+fr:Bonite crue
```
Comment (teolemon) [inline taxonomies/ingredients.txt:66441]:
```suggestion
en:raw bonito
fr:Bonite crue
```

File: taxonomies/ingredients.txt [hunk 17]
```diff
@@ -65290,7 +66812,43 @@ wikipedia:fr:https://fr.wikipedia.org/wiki/Pagellus
...
+
+<en:gilthead sea bream
+en:Raw wild gilthead seabream
+fr:Dorade royale sauvage crue, Daurade sauvage crue, Vraie daurade sauvage crue
+ciqual_food_code:en:26080
+ciqual_food_name:en:Gilthead seabream, raw, wild
+ciqual_food_name:fr:Dorade royale, ou daurade ou vraie daurade, crue, sauvage
+
+<en:gilthead sea bream
+en:Farmed wild gilthead seabream
```
Comment (teolemon) [inline taxonomies/ingredients.txt:66827]:
```suggestion
en:Farmed wild gilthead seabream, wild gilthead seabream from farming
```

File: taxonomies/ingredients.txt [hunk 18]
```diff
@@ -65520,10 +67090,15 @@ zh:虹鱒
...
 wikipedia:en:https://en.wikipedia.org/wiki/Rainbow_trout
 # ingredient/fr:truite-arc-en-ciel has 81 products in 3 languages @2018-12-30
-# structure:outer:role:synonym,  eg Truite arc-en-ciel (oncorhynchus mykiss)
 carbon_footprint_fr_foodges_ingredient:fr:Truite d'élevage, entière
 carbon_footprint_fr_foodges_value:fr:4.9
-# structure:inner:role:synonym,  eg:Truite (Oncorhynchus mykiss)
+
+<en:rainbow trout
+en:Rainbow trout raw farmed
+fr:Truite arc en ciel crue élevage
```
Comment (teolemon) [inline taxonomies/ingredients.txt:67098]:
```suggestion
en:Farmed rainbow trout, rainbow trout from farming
fr:Truite arc en ciel crue d'élevage
```

File: taxonomies/ingredients.txt [hunk 19]
```diff
@@ -65589,6 +67174,23 @@ en:turbot
 fi:piikkikampela
 fr:turbot
 wikipedia:en:https://en.wikipedia.org/wiki/Turbot
+ciqual_food_code:en:26174
+ciqual_food_name:en:Turbot, raw
+ciqual_food_name:fr:Turbot, cru
+
+<en:turbot
+en:Farmed turbot
```
Comment (teolemon) [inline taxonomies/ingredients.txt:67182]:
```suggestion
en:Farmed turbot, turbot from farming
```

File: taxonomies/ingredients.txt [hunk 20]
```diff
@@ -71546,6 +73284,15 @@ zh:蛋白
 wikidata:en:Q796758
 wikipedia:en:https://en.wikipedia.org/wiki/Egg_white
 # ingredient/fr:blanc-d-oeuf has 2748 products in 13 languages @2018-12-18
+ciqual_food_code:en:22001
+ciqual_food_name:en:Egg white, raw
+ciqual_food_name:fr:Oeuf, blanc (blanc d'oeuf), cru
+
+en:Egg white cooked
```
Comment (teolemon) [inline taxonomies/ingredients.txt:73291]:
```suggestion
en:cooked egg white, egg white cooked
```

File: taxonomies/ingredients_processing.txt [hunk 21]
```diff
@@ -1309,7 +1319,14 @@ hu:termesztett
...
 #we would need corresponding code changes so that NOVA 4 ingredient processing are used when computing NOVA groups.
- 
+
+en:non-hydrogenated
+de:ungehärtet
+es:no hidrogenada
+fr:non hydrogénée
+it:non idrogenato
+nl:niet gehydrogeneerde
+ptnâo hidrogenada
```
Comment (teolemon) [inline taxonomies/ingredients_processing.txt:1329]:
```suggestion
pt:nâo hidrogenada
```

File: taxonomies/ingredients.txt [hunk 22]
```diff
@@ -6134,17 +6140,39 @@ wikidata:en:Q57312618
...
+<en:unsalted butter
+en:82% fat unsalted Butter 
+fr:Beurre à 82% MG doux
+ciqual_food_code:en:16400
+ciqual_food_name:en:Butter, 82% fat, unsalted
+ciqual_food_name:fr:Beurre à 82% MG, doux
+
+<en:unsalted butter
+en:82% fat unsalted butter
+fr:Beurre à 82% MG doux
```
Comment (teolemon) [inline taxonomies/ingredients.txt:6152]:
```suggestion
en:82% fat unsalted Butter 
fr:Beurre doux à 82% MG
ciqual_food_code:en:16400
ciqual_food_name:en:Butter, 82% fat, unsalted
ciqual_food_name:fr:Beurre à 82% MG, doux

<en:unsalted butter
en:82% fat unsalted butter
fr:Beurre doux à 82% MG
```

File: taxonomies/ingredients.txt [hunk 23]
```diff
@@ -6134,17 +6140,39 @@ wikidata:en:Q57312618
...
 nl:halfgezouten boter
 wikipedia:fr:https://fr.wikipedia.org/wiki/Beurre_demi-sel
+description:en:BEURRE DEMI-SEL is a butter with a salt percentage between 0,5 % and 3 %
+description:fr:BEURRE DEMI-SEL - beurre avec une salinité moyenne de 0,5 % à 3 %
 # ingredient/fr:beurre-demi-sel has 115 products in french @2019-07-10
 
+<en:lightly salted butter
+en:Butter 80% fat lightly salted
+fr:Beurre à 80% MG demi-sel
+ciqual_food_code:en:16402
```
Comment (teolemon) [inline taxonomies/ingredients.txt:6171]:
```suggestion
<en:lightly salted butter
en:lightly salted butter with 80% fat
fr:Beurre demi-sel à 80% MG
ciqual_food_code:en:16402
```

**Comment by teolemon** [comment]
I realised we should probably be putting an origin property when the ingredient is a PGI.

File: taxonomies/ingredients.txt [hunk 24]
```diff
@@ -18010,20 +19460,24 @@ es:aceite orujo de oliva
...
 <en:vegetable oil and fat
 en:rapeseed oil, rapeseed fat, rapeseed oil and fat
 cs:řepkový olej, olej řepkový
 da:rapsolie
 de:Rapsöl, Rapsol
+eo:rapoleo
 es:aceite de nabina, aceite vegetal de nabina
 et:rapsiõli, rapsiseemneõli, rüpsiõli
 fi:rapsiöljy, rapsirasva, rapsiöljy ja -rasva
-fr:Huile de colza ou de navette
```
Comment (teolemon) [inline taxonomies/ingredients.txt:18021]: @stephanegigandet I don't remember about that one

**Comment by aleene** [comment]
Note that it is not finished yet. Much more to be added. But I use the opportunity to clean up a bit as well.

**Comment by stephanegigandet** [comment]
> I do not understand why the test (in tags.t) fails. I see no relation with any changes I made yesterday.

@aleene: in the tags.t structure, the tags are ordered by "level". And the level of a tag is the maximum number of child levels it has below it. So if you added some children to some tags, that might have changed the level of their parents. That's fine, but the test expected result needs to be updated.

**Comment by aleene** [comment]
Difficult test. Only the test result can tell me if something is changed, but it does not tell where the change took place. So if I made an error somewhere, I get no info where that happened.

**Pull request merged**